---
title: "antibiotic_exposure_analysis_tidy"
author: "Ashley Bell"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---
# Setup my R session
```{r setup, echo = FALSE}
#only affects subsequent chunks
knitr::opts_knit$set(root.dir = "/Users/agb214/OneDrive - University of Exeter/PhD/Antibiotics_exposure/")
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
setwd("/Users/agb214/OneDrive - University of Exeter/PhD/Antibiotics_exposure/")
```

# Load libraries
```{r libraries}
library(tidyverse)
library(lmerTest)
library(picante)
library(phyloseq)
library(decontam)
library(microViz)
library(RColorBrewer)
library(gridExtra)
library(ggpubr)
library(ape)
library(ggh4x)
library(patchwork)
library(cowplot)
library(reshape2)
library(furrr)
library(progressr)
library(GUniFrac)
library(ggalluvial)

source("/Users/agb214/OneDrive - University of Exeter/PhD/R_files/pairwise.adonis2.R")

# Ashley preferences
set.seed(1234)
options(scipen=999) # no scientific notation for integers
`%!in%` = Negate(`%in%`) # not in function
```

# Define my global colour palette 
```{r}
#define colours for tanks
blue <- c("#01377d","#009dd1", "#48AAAD")# control
yellow <- c("#F58216", "#FFBF00", "#FBE106") # low dose
red <- c("#760F13","#d61e13","#f76e6d") # high dose

clrs.tanks <- c(rep(c(red, yellow, blue), 2), red[2],yellow[2],blue[2])
names(clrs.tanks) <- c("1A","1B","1C","2A","2B","2C","3A","3B","3C", 
                       "HighA","HighB","HighC",
                       "LowA","LowB","LowC",
                       "ControlA","ControlB","ControlC",
                       "High","Low","Control")
#show_col(c(blue, red, yellow))
```

# Import phyloseq obj and data curation
```{r}
metadata <- read.csv("data/phyloseq_metadata.csv")
convert <- read_csv("data/well.position.to.fwd.primer.csv", col_types = c("ff"))

ps <- readRDS("data/phyloseq_taxa_tree.rds")

metadata2 <- metadata %>%
  select(Plate.no, well) %>%
  rownames_to_column(var = "rowname") %>%
  mutate(suffix = ".filt.fwd.fq.gz", 
         seq_plate = case_when(Plate.no == 1 ~ "11143_AP1_S114_",
                               Plate.no == 2 ~ "11143_AP2_S115_",
                               Plate.no == 3 ~ "11143_AP3_S116_",
                               Plate.no == 4 ~ "11143_AP4_S117_",
                               Plate.no == 5 ~ "11143_AP5_S118_")) %>%
  rename(Well.position = well) %>%
  left_join(convert, by = "Well.position") %>%
  mutate(phyloseq.sample.name = paste0(seq_plate, fwd.primer.name, suffix), .keep = "unused") %>%
  filter(phyloseq.sample.name != "NANA.filt.fwd.fq.gz") %>%
  select(-Plate.no, -Well.position) %>%
  left_join(rownames_to_column(metadata, var = "rowname"), by = "rowname") %>%
  select(-rowname) %>%
  column_to_rownames(var = "phyloseq.sample.name")

sample_data(ps) <- metadata2
# checked three random rows, metadata appears correct

ps <- tax_fix(ps)
```

### We should probably root our phylotree for future functions
See https://github.com/joey711/phyloseq/issues/597
```{r}
pick_new_outgroup <- function(tree.unrooted){
    require("magrittr")
    require("data.table")
    require("ape") # ape::Ntip
    # tablify parts of tree that we need.
    treeDT <- 
      cbind(
        data.table(tree.unrooted$edge),
        data.table(length = tree.unrooted$edge.length)
      )[1:Ntip(tree.unrooted)] %>% 
      cbind(data.table(id = tree.unrooted$tip.label))
    # Take the longest terminal branch as outgroup
    new.outgroup <- treeDT[which.max(length)]$id
    return(new.outgroup)
  }

new.outgroup = pick_new_outgroup(phy_tree(ps))
rootedTree = ape::root(phy_tree(ps), outgroup=new.outgroup, resolve.root=TRUE)

phy_tree(ps) <- phy_tree(rootedTree)
```

# Supplementary Figure 3
## Positive controls
### What do my positive controls look like?
 ZymoBIOMICS Microbial Community Standard (Zymo; D6300) (contains cells) in all Postive samples.
```{r mock community comparison}
zymo.mock <- data.frame(sample = rep("Zymo", 8), 
                        rel.abundance = c(4.2,10.1,10.4,18.4,9.9,15.5,14.1,17.4),
                        Species = c("Pseudomonas aeruginosa","Escherichia coli","Salmonella enterica","Lactobacillus fermentum","Enterococcus faecalis","Staphylococcus aureus","Listeria monocytogenes","Bacillus subtilis"),
                        Genus = c("Pseudomonas","Escherichia","Salmonella","Lactobacillus","Enterococcus","Staphylococcus","Listeria","Bacillus"),
                        Genus2 = c("Pseudomonas","Escherichia-Shigella","Salmonella","Limosilactobacillus","Enterococcus","Staphylococcus","Listeria","Bacillus"))

zymo.mock <- zymo.mock[order(zymo.mock$Genus2), ]

clrs <- rep(brewer.pal(8, "Set1"), 3)
names(clrs) <- c(zymo.mock$Genus, zymo.mock$Species, zymo.mock$Genus2)


pos.plot <- ps %>%
  ps_filter(Is.control == "Positive") %>%
  tax_transform("compositional", rank = "Genus") %>%
  comp_barplot(tax_level = "Genus", tax_order = sort(zymo.mock$Genus2, decreasing = T)) +
  scale_fill_manual(values = clrs) +
  guides(fill = guide_legend(reverse=T)) +
  ylab("Relative Abundance") +
  theme(axis.text.x = element_blank())

zymo.p <- ggplot(zymo.mock, aes(x = sample, y = rel.abundance/100, fill = Species)) +
  geom_bar(stat= "identity", colour="black") +
  scale_fill_brewer(palette = "Paired") +
  theme_minimal() +
  scale_fill_manual(values = clrs) +
  ylab("Relative Abundance") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.title.x = element_blank())
  
cowplot::plot_grid(zymo.p, pos.plot, nrow = 1, labels = "AUTO", align = "v", rel_widths = c(1,1.75))
```

# Run Decontam to remove contaminate ASVs
## Using the decontam frequency method, what contams are there?
So for the frequency method, you're meant to use the DNA concentration of samples before sequencing. This doesn't make sense because sequencing would equimolar pool your samples to get the same sequencing read depth. Anyways, we we equimolar pooled our samples to 10ng/uL. We could use either the DNA concentration before PCR (doesn't make much sense as it contains non-bacterial DNA) concentration after PCR clean-up but before equimolar pooling (Doesn't make sense because we increase/decrease each sample's concentration before sequencing). In addition our negative samples have no DNA concentration as it was below the limits of detection (unsurprising as we practice sterile techniques). Phyloseq requires positive values, so I define NA or 0 DNA conc as half of the lowest DNA concentration found in all samples (MicroViz "halfmin" solution). I've also defined the sample concentration as the sample_sums(ps) as this is the number of reads I got back from sequencing and the most representative of each samples "DNA concentration" before sequencing.

To avoid messing up my data, this is now the "quant_reading" sample data column
```{r decontam frequencies}
sample_data(ps)$is.neg <- sample_data(ps)$Is.control %in% c("Negative", "NTC")
sample_data(ps)$quant_reading <- sample_sums(ps)

sample_data(ps) %>%
  data.frame %>%
  filter(is.neg == T) %>%
  with(., table(quant_reading))

# calculate contam
contamdf.freq <- isContaminant(ps, method="frequency", conc="quant_reading", threshold = 0.05)

# how many contam?
table(contamdf.freq$contaminant)

# 0.5 is too overzealous with detection as a lot of dots are outside of the expected line 0.1 default seems to be best
plot_frequency(ps, taxa_names(ps)[which(contamdf.freq$contaminant)[1:25]], conc="quant_reading") +
    xlab("Sample sums from each sample")
```

0.5, 0.25 and 0.05 was too much, 0.05 seemed best

## What is the decontam prevalence method?
```{r decontam prevalence method}

contamdf.prev <- isContaminant(ps, method="prevalence", neg="is.neg", threshold=0.25)

# how many ASVs are contams
table(contamdf.prev$contaminant)

# plot identified contams by frequency and DNA
ps.pa <- transform_sample_counts(ps, function(abund) 1*(abund>0))
ps.pa.neg <- prune_samples(sample_data(ps.pa)$Is.control %in% c("Negative Control", "NTC"), ps.pa)
ps.pa.pos <- prune_samples(sample_data(ps.pa)$Is.control %!in% c("Negative Control", "NTC"), ps.pa)

# Make data.frame of prevalence in positive and negative samples
df.pa <- data.frame(pa.pos=taxa_sums(ps.pa.pos), pa.neg=taxa_sums(ps.pa.neg),
                      contaminant=contamdf.prev$contaminant)

# 0.5 removes a lot of taxa in the plot_frequency plot that don't fit the trend. went for 0.25 instead
ggplot(data=df.pa, aes(x=pa.neg, y=pa.pos, color=contaminant)) + geom_point() +
  xlab("Prevalence (Negative Controls)") + ylab("Prevalence (True Samples)")

plot_frequency(ps, taxa_names(ps)[sample(which(contamdf.prev$contaminant),25)], conc="quant_reading") +
    xlab("DNA Concentration before PCR")
```

## Are any of the combination methods better?
"combined" method seemed not to remove much 
"both" method removed almost nothing
so I went with the "either" method
```{r combination methods}
contamdf.comb <- isContaminant(ps, con = "quant_reading", method="both", neg="is.neg", threshold=c(0.05, 0.25))

table(contamdf.comb$contaminant)

contamdf.either <- isContaminant(ps, con = "quant_reading", method="either", neg="is.neg", threshold=c(0.05, 0.25))

table(contamdf.either$contaminant)

plot_frequency(ps, taxa_names(ps)[sample(which(contamdf.either$contaminant),25)], conc="quant_reading") +
    xlab("DNA Concentration before PCR")
```

## Remove non-bacterial sequences and contaminants
```{r remove non-bacteria}

### happy with that, prune contams and neg samples
ps.noncontam <- prune_taxa(!contamdf.either$contaminant, ps)
ps.noncontam

ps.noncontam.noeuk <- subset_taxa(ps.noncontam, Family != "Mitochondria")
ps.noncontam.noeuk <- subset_taxa(ps.noncontam.noeuk, Order != "Chloroplast")
ps.noncontam.noeuk <- subset_taxa(ps.noncontam.noeuk, Kingdom != "Eukaryota")
ps.noncontam.noeuk

# Lets remove the pos/neg/ntc samples as they aren't needed for further analyse
ps <- ps_filter(ps.noncontam.noeuk, Is.control %!in% c("Positive", "Negative", "NTC"))

hist(sample_sums(ps), breaks = length(sample_sums(ps)))
quantile(sample_sums(ps), probs = c(0, 0.01, 0.05, 0.95, 0.99, 1))
# my smallest sample has an abundance of 4 which makes analysis difficult. If remove the bottom 1% of samples, how many samples does this remove and what is the cutoff?

length(sample_sums(ps))
bottom.perc <- sample_sums(ps)[sample_sums(ps) < quantile(sample_sums(ps), probs = 0.0025)]
bottom.perc
length(bottom.perc) # 5 samples within bottom 5%
# I want to two samples at 57 and 4 abundance as the next one up is 9491 abundance. I'll just say I removed any samples less than any number between 100 and 7500. Picked 5000 as its easiest.

ps <- ps_filter(ps, sample_sums(ps) >= 100)
meta <- sample_data(ps) %>% data.frame()
saveRDS(ps, "data/pruned_pyloseq.rds")
```

# Samples are clean, we can start analysis


# Supplementary Figure 4
## Antibiotics dosages
```{r}
#####
# What are the antibiotic concentrations through time
AB <- read_csv("data/water.chemistry_sample.tubes.csv") %>%
  mutate(across(.cols = everything(), ~str_replace(., "< MQL|< MDL|< IQL|< IDL|#N/A|No ISTD", "0"))) %>%
  replace(is.na(.), "0") # replace undetected antibitoics as concentration 0. R can't work with columns of different types

# Get cols relevant to antibiotics
AB.colnames <- colnames(AB)[grepl("µg/L", colnames(AB))]

# Make these columns numeric
AB[AB.colnames] <- sapply((AB[AB.colnames]), as.numeric)

# Just get recorded antibiotic concentrations relevant to this study (remove metabolites)
AB.filt <- AB %>%
  select(Days.post.exposure.start, Tank, AB.concentration.group,
         all_of(AB.colnames)) %>%
  filter(AB.concentration.group != "Matrix QC") %>%
  mutate(Days.post.exposure.start = as.numeric(Days.post.exposure.start)) %>%
  pivot_longer(cols = all_of(AB.colnames),
               names_to = "Antibiotic_Group", values_to = "Antibiotic_Concentration") %>%
  separate(Antibiotic_Group, c("target.actual","target.antibiotic", NA), sep = "\\.") %>%
  filter(target.antibiotic != "desethylene_ciprofloxacin", target.antibiotic != "desmethyl_clarithromycin") %>%
  mutate(AB.concentration.group = if_else(AB.concentration.group == "high.dose", "High dose exposure", 
                                          if_else(AB.concentration.group == "low.dose", "Low dose exposure", "Negative control")),
         target.antibiotic = str_to_title(target.antibiotic))

AB.actual <- AB.filt %>% filter(target.actual == "actual")
AB.target <- AB.filt %>% filter(target.actual == "target")

# Make a fake column where anbitoics at day 1 is 0 to plot target concentration only
AB.target.day0 <- AB.target %>% 
  filter(Days.post.exposure.start == 1) %>%
  mutate(Days.post.exposure.start = -0.1, 
         Antibiotic_Concentration = 0)

AB.target <- rbind(AB.target,AB.target.day0)

# Plot  
ggplot(data = AB.actual, aes(x = Days.post.exposure.start, 
                             y = Antibiotic_Concentration,
                             colour = target.actual)) +
  geom_point() +
  facet_grid2(AB.concentration.group ~ target.antibiotic, 
              scales = "free_y", independent = "y") +
  geom_smooth(se = F) +
  scale_y_continuous(n.breaks = 10) +
  ylab("Antibiotic Concentration (µg/L)") +
  scale_colour_brewer(palette = "Set2") +
  theme_bw() +
  geom_line(data = AB.target, linewidth = 1) +
  xlab("Days post exposure start") +
  coord_cartesian(ylim = c(0, NA))

# Get some stats on the actual antibiotic concentrations observed during exposure
AB.filt %>%
  filter(target.actual == "actual") %>%
  group_by(AB.concentration.group, Days.post.exposure.start, target.antibiotic) %>%
  summarise(mean(Antibiotic_Concentration), .groups = "drop") %>%
  filter(Days.post.exposure.start %in% c(7))
```

# Create Rarified ASV tables
```{r}
### This method rarefaction the OTU table. But looking at Pat Schloss's work, he rarefies then calculates a diversity metric from that

# ps <- readRDS("data/pruned_pyloseq.rds")

# calculate rarefaction instead of rarifying
# this takes around 16 minutes so saved output as RDS
# Convert sample names to numbers to make more memory efficient
# sample_name_mapping <- sample_data(ps) %>%
#   data.frame() %>%
#   rownames_to_column(var = "Sample") %>%
#   select(Sample) %>%
#   unique() %>%
#   mutate(sample_number = as.numeric(factor(Sample, levels = unique(.$Sample))))
# 
# Get ASV_table, convert to long format, remove rows with counts of 0, then instead make the number of unique ASVs that occur in a sample as a weighting rather than raw counts (similar to relative abundance). This make memory requirements and computation time lower
# ASV.tab <- otu_table(ps) %>%
#   as.data.frame() %>%
#   rownames_to_column(var = "Sample") %>%
#   pivot_longer(-Sample, names_to = "ASV", values_to = "count") %>%
#   filter(count > 0) %>%
#   group_by(Sample) %>%
#   mutate(weighting = count / sum(count)) %>%
#   select(-count) %>%
#   ungroup() %>%
#   left_join(sample_name_mapping, by = "Sample") %>%
#   select(-Sample) %>%
#   tibble() %>%
#   group_by(sample_number)
# 
# Define my minimum sampling size as 95% of my smallest sample so I get some variation in my smallest sample
# sample_size <- round(min(sample_sums(ps)) * 0.95)
# 
# resample with replacement because I used weightings rather than raw counts. The weights affect how likely I am to pick that ASV from that sample
# gc()
# counter <- 0
# start <- Sys.time()
# rarefaction <- bind_rows(lapply(1:1000, function(i) {
#   print(counter <<- counter + 1);
#     slice_sample(ASV.tab,
#       n = sample_size,
#       replace = T,
#       weight_by = weighting) %>%
#     select(-weighting) %>%
#     count(ASV) %>%
#     mutate(iteration = counter)
#   })
# )
# (time <- round(Sys.time() - start, digits = 2))
# gc()
# 
# Convert my sample numbers back to names
# rarefaction <- rarefaction %>%
#   ungroup() %>%
#   left_join(sample_name_mapping, by = "sample_number") %>%
#   select(-sample_number)
# 
# write_csv(rarefaction, file = "data/rarefaction.csv")
```

# Figure 2 and Supplementary table 3
## Alpha diversity
```{r}
ps <- readRDS("data/pruned_pyloseq.rds")
rarefaction <- read_csv(file = "data/rarefaction.csv") %>%
  nest(ASV_tab = c("Sample", "ASV", "n")) %>%
  mutate(data = map(data, function(x)
    pivot_wider(x, id_cols = "Sample", names_from = "ASV",
                    values_from = "n", values_fill = 0))) %>%
  ungroup()

# Get Faith's PD
# PD needs a phylogenetic tree. This function prunes the PS.obj tree to just taxa found in each ASV_table
get_pruned_tree <- function(PS_TREE, ASV_NAMES) {
  ape::drop.tip(PS_TREE, PS_TREE$tip.label[!(PS_TREE$tip.label %in% ASV_NAMES)])
}

library(furrr)  #multithread. This takes awhile
gc()
plan(multisession, workers = 2)
start <- Sys.time()
alpha.div <- rarefaction %>%
  mutate(Faith.PD = future_map(ASV_tab, function(df)
    picante::pd(samp = df, tree = get_pruned_tree(phy_tree(ps), colnames(df)), include.root = T) %>%
      data.frame() %>%
      rownames_to_column(var = "Sample"),
    .options = furrr_options(seed=T)))
(time <- round(Sys.time() - start, digits = 2))
plan(sequential)

# Shannon and Chao1
# this is comparative quick so done without multithread
alpha.div <- rarefaction %>%
  group_by(iteration) %>%
  nest(ASV_tab = c()) %>%
  mutate(data = map(data, function(x)
    pivot_wider(x, id_cols = "Sample", names_from = "ASV",
                    values_from = "n", values_fill = 0) %>%
      column_to_rownames(var = "Sample"))) %>%
  ungroup() %>%
  rename(ASV_tab = data) %>%
  mutate(Shannon = map(ASV_tab, function(df)
    diversity(df, index = "shannon") %>%
      data.frame() %>%
      rownames_to_column(var = "Sample")),
         Chao1 = map(ASV_tab, function(df)
    t(estimateR(df, index = "chao")) %>%
      data.frame() %>%
      rownames_to_column(var = "Sample")))

Shannon <- alpha.div %>%
  select(iteration, Shannon) %>%
  unnest(cols = c(Shannon)) %>%
  group_by(Sample) %>%
  summarise(Shannon = mean(`.`))

Chao1 <- alpha.div %>%
  select(iteration, Chao1) %>%
  unnest(cols = c(Chao1)) %>%
  group_by(Sample) %>%
  summarise(Chao1 = mean(S.chao1))

Faith.PD <- alpha.div %>%
  select(iteration, Faith.PD) %>%
  unnest(cols = c(Faith.PD)) %>%
  group_by(Sample) %>%
  summarise(PD = mean(PD))

alpha.div.mean <- Shannon %>%
  full_join(Chao1, by = "Sample") %>%
  full_join(Faith.PD, by = "Sample")

#write_csv(alpha.div.mean, "alpha.diversity.mean.csv")

metadata <- samdat_tbl(ps) %>%
  rename(Sample = .sample_name)

alpha.div <- read_csv("data/alpha.diversity.mean.csv") %>%
  left_join(metadata, by = "Sample") %>%
  mutate(tank = as.factor(paste0(treatment, replicate))) %>%
  tibble()

 alpha.div %>%
  pivot_longer(cols = c("Chao1", "Shannon", "PD"), names_to = "alpha.div.name", values_to = "alpha.div.value") %>%
  mutate(tank = paste0(treatment, replicate)) %>%
  ggplot(aes(x = Days.post.exposure.start, y = alpha.div.value)) +
  annotate("rect", xmin = 0, xmax = 14, ymin = -Inf, ymax = Inf, 
           alpha=0.2, fill="black") +
  geom_jitter(aes(fill = tank, color = tank), width = 0.4, alpha = 0.25) +
  #geom_boxplot(aes(x = as.numeric(levels(Days.post.exposure.start)))) +
  geom_line(stat = "smooth", aes(group = tank, color = tank), 
            se = F, method = "loess", alpha = 0.2) +
  geom_line(stat = "smooth", aes(group = treatment, color = treatment), 
            se = F, method = "loess", linewidth = 1) +
  facet_grid2(sample.type~alpha.div.name, scales = "free", independent = "all") +
  scale_fill_manual(values = clrs.tanks) +
  scale_color_manual(values = clrs.tanks) +
  theme_bw() +
  xlab("Days post exposure start") +
  ylab("Alpha diversity")
```

# Alpha diversity stats
```{r}
ps <- readRDS("data/pruned_pyloseq.rds")
metadata <- samdat_tbl(ps) %>%
  rename(Sample = .sample_name)


alpha.div.stats <- alpha.div %>%
  select(-ASV_tab) %>%
  mutate(alpha.div = map2(Shannon, Chao1, ~full_join(.x, .y, by = "Sample")), 
         alpha.div = map2(alpha.div, Faith.PD, ~full_join(.x, .y, by = "Sample"))) %>%
  select(iteration, alpha.div) %>%
  unnest(cols = alpha.div) %>%
  select(., iteration, Sample, `.`, S.chao1, PD) %>%
  rename(., Shannon = `.`,
         Chao1 = S.chao1) %>%
  left_join(metadata, by = "Sample") %>%
  pivot_longer(cols = c(Shannon, Chao1, PD), names_to = "alpha.div.name", values_to = "alpha.div.value")

# Define my model and how to subset the data
lm.test <- function(TP1, TP2, metric, sam.type, iter){
  subsetted.data <- alpha.div.stats %>%
      filter(Days.post.exposure.start %in% c(TP1, TP2),
      alpha.div.name == metric,
      sample.type == sam.type,
      iteration == iter)
    
    # This allows for treatment to have a random slope while noting tank is nested in treatment group. Can't include PIT.tag as only one replicate per time point = can't be grouped
    model <- lmerTest::lmer(alpha.div.value ~ treatment +
                           (1 + treatment|treatment/tank) +
                           (1|Days.post.exposure.start), 
                           data = subsetted.data)
   
    model.clean <- summary(model)$coefficients %>%
      as.data.frame() %>%
      rownames_to_column(var = "test") %>%
      mutate(TP = paste0(TP1, "_vs_", TP2),
            alpha.div = metric, 
            sample.type = sam.type,
            iteration = iter)
    
    return(model.clean)
}

# comparisons to make
comp <- c(-2,7,14,21,28)
comp <- data.frame(TP1 = comp[1:length(comp)-1],
                   TP2 = comp[-1])

# This allows for treatment to have a random slope while noting tank is nested in treatment group. Can't include PIT.tag as only one replicate per time point = can't be grouped
data <- data.frame()
for (iter in 1:max(alpha.div.stats$iteration)) {
  print(iter)
  for (metric in list("Shannon", "Chao1", "PD")) {
    for (tests in 1:nrow(comp)) {
      for (sample.type in c("Filter", "Swab")) {
        model.data <- lm.test(comp$TP1[tests], comp$TP2[tests],
                              metric, sample.type, iter)
        data <- rbind(model.data, data)
      }
    }
  }
}

#write_csv(data, "data/alpha.div.stats.csv")

stats.table <- data %>%
  filter(test %!in% c("(Intercept)")) %>%
  group_by(iteration) %>%
  mutate(p.adj = p.adjust(`Pr(>|t|)`, method = "BH")) %>%
  ungroup() %>%
  rename(p.unadj = `Pr(>|t|)`) %>%
  group_by(alpha.div, sample.type, TP, test) %>%
  summarise(across(where(is.double), mean)) %>%
  mutate(sig.stat = if_else(p.adj <= 0.01, "***", 
                            if_else(p.adj <= 0.01, "**", 
                                    if_else(p.adj <= 0.05, "*",
                                            if_else(p.adj > 0.05, "N.S.", ""))))) %>%
  select(-`Std. Error`, -df, -`t value`) %>%
  arrange(sample.type, TP, alpha.div, sig.stat)

stats.table %>%
  rename(`Alpha diversity metric` = alpha.div,
         `Sample type` = sample.type,
         `Treatment group` = test,
         Significance = sig.stat,
         `B.H. adjusted p value` = p.adj) %>%
  separate(TP, into = c("Timepoint 1", "Timepoint 2"), sep = "_vs_") %>%
  mutate(`Treatment group` = str_remove(`Treatment group`, "treatment")) %>%
  select(-p.unadj) %>%
  mutate(across(where(is.double), ~ round(., digits = 2))) %>%
  mutate(`B.H. adjusted p value` = if_else(`B.H. adjusted p value` == 0,
                                           as.character("p < 0.01"), as.character(`B.H. adjusted p value`)))

# no measures significant
```

# Figure 3 and Supplementary table 4 and 5
## Beta Diversity
```{r}
ps <- readRDS("data/pruned_pyloseq.rds")
get_pruned_tree <- function(PS_TREE, ASV_NAMES) {
  ape::drop.tip(PS_TREE, PS_TREE$tip.label[!(PS_TREE$tip.label %in% ASV_NAMES)])
}

rarefaction <- read_csv("data/rarefaction.csv") %>%
  nest(ASV_tab = c("Sample", "ASV", "n")) %>%
  mutate(ASV_tab = map(ASV_tab, function(df)
    pivot_wider(df, id_cols = "Sample", names_from = "ASV",
                values_from = "n", values_fill = 0) %>%
      column_to_rownames(var = "Sample"))) %>%
    mutate(rclr = map(ASV_tab,
                      function(df) vegan::vegdist(df, method = "robust.aitchison")),
           BC = map(ASV_tab,
                    function(df) vegan::vegdist(df, method = "bray")),
           Jacc = map(ASV_tab,
                      function(df) vegan::vegdist(df, method = "jaccard",
                                                  binary = T)),
           unifracs = map(ASV_tab,
                      function(df) GUniFrac::GUniFrac(ASV_tab, phy_tree(get_pruned_tree(phy_tree(ps), colnames(df))), alpha = c(0.5, 1))$unifracs),
           WUniFrac_0.5 = map(unifracs,
                      function(df) as.dist(df[, , "d_0.5"])),
           WUniFrac = map(unifracs,
                      function(df) as.dist(df[, , "d_1"]))
           )

#saveRDS(rarefaction, "data/rarefaction.rds")

BC.agg <- Reduce(`+`, rarefaction$BC) / length(rarefaction$BC)
rclr.agg <- Reduce(`+`, rarefaction$rclr) / length(rarefaction$rclr)
Jacc.agg <- Reduce(`+`, rarefaction$Jacc) / length(rarefaction$Jacc)
wUF5 <- Reduce(`+`, WUniFrac_0.5)/ length(WUniFrac_0.5)
wUF <- Reduce(`+`, WUniFrac)/ length(WUniFrac)
rm(rarefaction)
gc()

# So we can swap in our rarified (rarefaction 1000 times) matrix after filtering which samples we want to preseved the metadata created using MicroViz's transformation
# Note: We calculate stats on the dissim matrix not the ordination (NMDS) therefore don't need to bootstrap 1000 NMDS plots - its just for visualisation.

bc.ord <- function(sam.type){
  bc.matrix <- ps %>%
    ps_filter(sample.type == sam.type) %>%
    tax_fix() %>%
    tax_transform(trans = "identity", rank = NA) %>%
    dist_calc("bray")
  
  filt.sams <- dist_get(bc.matrix) %>%
    as.matrix() %>%
    rownames()
  
  bc.matrix@dist <- as.matrix(BC.agg)[filt.sams, filt.sams] %>%
    as.dist()
  
  return(bc.matrix)
}

ord.plot <- function(ordination){
  ordination %>%
  ord_calc(method = "NMDS") %>%
  ord_plot(colour = "treatment") +
  ggpubr::stat_chull(aes(fill = treatment, color = treatment), alpha = 0.1, geom = "polygon") +
  #ggpubr::stat_stars(aes(color = treatment)) +
  scale_fill_manual(values = clrs.tanks, aesthetics = c("colour", "fill")) +
  ggh4x::facet_grid2(~Days.post.exposure.start, scales = "free", independent = "all") +
  theme_void()
}

bc.vs.ctrl <- function(sam.type){
  ps2 <- bc.ord(sam.type) # get ordination from bc.ord function but only run once
  
  ps2 %>%
  dist_get() %>% # get BC matrix from PS Extra obj
  as.matrix() %>% # as matrix
  reshape2::melt() %>% # pivot longer format
  left_join(samdat_tbl(ps2), by = join_by(Var1 == .sample_name)) %>% # get metadata
  group_by(Days.post.exposure.start) %>% # groupby day samples where taken
  filter(Var2 %in% Var1) %>% # remove comparisons across sample days
  ungroup() %>% # ungroup
  select(Var1, treatment, Var2, value) %>% # only keep dis matrix and comparison
  left_join(samdat_tbl(ps2), by = join_by(Var2 == .sample_name)) %>% # join by var 2 instead
  select(Var1, treatment.x, Var2, treatment.y, Days.post.exposure.start, value) %>%
  filter(treatment.x == "Control" | treatment.y == "Control") %>% # remove comparisons not 
  unite("treatment", treatment.x, treatment.y, na.rm = T) %>% # combine treatments into one column
  group_by(treatment, Days.post.exposure.start) %>%
  summarise(value = median(value), .groups = "drop") # Just keep median data
}

diff.plot <- function(BC.vs.ctrl.df) {
  BC.vs.ctrl.df %>%
    mutate(treatment = str_replace(treatment, ".*High.*", "High"),
           treatment = str_replace(treatment, ".*Low.*", "Low"),
           treatment = str_replace(treatment, "Control_Control", "Control")) %>%
  ggplot(aes(x = Days.post.exposure.start, y = value, colour = treatment)) +
  geom_line(linewidth = 1) +
  annotate("rect", xmin = 3, xmax = 16, ymin = -Inf, ymax = Inf, 
         alpha=0.2, fill="black") +
  scale_fill_manual(values = clrs.tanks, aesthetics = c("colour", "fill")) +
  theme_minimal() +
  theme(legend.position = "none", panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.title.y=element_blank())
}

bc.ord.swab <- bc.ord("Swab")
bc.ord.swab.plot <- ord.plot(bc.ord.swab)

diff.swab <- bc.vs.ctrl("Swab")
diff.swab.plot <- diff.plot(diff.swab)

bc.ord.filt <- bc.ord("Filter")
bc.ord.filt.plot <- ord.plot(bc.ord.filt)

diff.filt <- bc.vs.ctrl("Filter")
diff.filt.plot <- diff.plot(diff.filt)

bc.swab.plots <- diff.swab.plot / bc.ord.swab.plot +
  patchwork::plot_layout(heights = c(1, 5)) + 
  patchwork::plot_annotation(tag_levels = "A") +
  guides(color=guide_legend("Treatment"),
         fill=guide_legend("Treatment"))

 bc.filt.plots<- diff.filt.plot / bc.ord.filt.plot + 
    patchwork::plot_layout(heights = c(1, 5)) + 
    patchwork::plot_annotation(tag_levels = list(c("C", "D"))) +
  guides(color=guide_legend("Treatment"),
         fill=guide_legend("Treatment"))

```

# Table 4 & 5
## Beta diversity stats
```{r}
UF5 <- read_csv("data/combined_UF_0.5_stats.table.csv") %>%
  format_table(.) %>%
  mutate(dist = "Weighted UniFrac 0.5")
UF <- read_csv("data/combined_UF_stats.table.csv") %>%
  format_table(.) %>%
  mutate(dist = "Weighted UniFrac")
BC <- read_csv("data/combined_BC_stats.table.csv") %>%
  format_table(.)%>%
  mutate(dist = "Bray-Curtis")
Jacc <- read_csv("data/combined_Jacc_stats.table.csv") %>%
  format_table(.)%>%
  mutate(dist = "Binary Jaccard")
rclr <- read_csv("data/combined_rclr_stats.table.csv") %>%
  format_table(.)%>%
  mutate(dist = "Aitchisons")

swab.stats.table <- BC %>% filter(type == "Swab")
filt.stats.table <- BC %>% filter(type == "Filter")


### SEE dissim_stats_ISCA_scripts.R for how I got to here

format_table <- function(STATS_TABLE) {
    STATS_TABLE %>%
    select(-filename, -significant) %>%
    group_by(pair1, pair2, Timepoint, type) %>%
    summarise(across(R2.PA:R2.PA.perc, mean), .groups = "drop") %>%
    mutate(p.adj.PA = round(p.adjust(p.val.PA, "BH"), digits = 3),
           R2.PA.perc = round(R2.PA, digits = 4)*100,
           significant = if_else(p.adj.bdisp <= 0.05, "bdisp_sig", 
                                 if_else(p.adj.PA <= 0.001, "***",
                                         if_else(p.adj.PA <= 0.01, "**",
                                                 if_else(p.adj.PA <= 0.05, "*",
                                                         if_else(p.adj.PA > 0.05, "N.S.", "")))))) %>%
  arrange(Timepoint, pair1, pair2) %>%
  select(pair1, pair2, type, Timepoint, p.adj.PA, R2.PA.perc, significant)
}

bind_rows(UF5, UF, BC, Jacc, rclr) %>%
  filter(pair2 == "Control") %>% # pair2 is always control
  ggplot(aes(x = Timepoint, y = R2.PA.perc, group = dist)) +
   annotate("rect", xmin = 3, xmax = 16, ymin = -Inf, ymax = Inf, 
         alpha=0.25, fill="black") +
  geom_line(linewidth = 2, aes(colour = dist)) +
  facet_wrap(type~pair1) +
  scale_fill_brewer(palette = "Set2", aesthetics = c("colour","fill")) +
  theme_minimal() +
  xlab("Days post exposure start") +
  ylab("Average PERMANOVA R² (%)") +
  labs(color="Dissimilarity Matrix")
```

# Supplementary Figure 6 and 7, Supplementary table 7
## Differential abundance
```{r}
# What taxa are enriched
ps <- readRDS("data/combined_ASV_PS.rds")

rarefaction <- readRDS("data/rarefaction.rds") %>%
  select(iteration, ASV_tab)

gc()

comp <- c(-2,7,14,21,28)
comp <- expand.grid(TP1 = comp, TP2 = comp)
comp <- comp[comp$TP1 != comp$TP2 & comp$TP1 < comp$TP2, ] %>%
  mutate(Pre_treatment = if_else(TP1 == -2 | TP2 == -2, T, F),
         Exposure = if_else(TP1 %in% c(7,14) | TP2 %in% c(7,14), T, F),
         Depuration = if_else(TP1 %in% c(21,28) | TP2 %in% c(21,28), T, F)) %>%
  pivot_longer(cols = c(Pre_treatment, Exposure, Depuration)) %>%
  filter(value != F) %>%
  select(-value) %>%
  group_by(TP1, TP2) %>%
  reframe(comparison = paste0(first(name),"_vs_",last(name)))

diff.abund <- tibble()
start <- Sys.time()
for (ITER in sample(1:1000, 1, replace = F)){
  otu_table(ps) <- otu_table(rarefaction$ASV_tab[[ITER]], taxa_are_rows = F)
  for (TEST in 1:nrow(comp)) {
    for (SAM.TYPE in c("Swab", "Filter")) {
      for (TREATMENT in c("Low", "High")) {
        if (SAM.TYPE == "Swab") {
          diff.abund <- bind_rows(diff.abund,
                              ps %>%
            tax_fix() %>%
            ps_filter(sample.type == SAM.TYPE,
                      Days.post.exposure.start %in% c(comp$TP1[TEST], comp$TP2[TEST]),
                      treatment %in% c("Control", TREATMENT)) %>%
            tax_agg(rank = "Genus") %>%
            tax_prepend_ranks() %>%
            tax_filter(min_sample_abundance = 30, tax_level = "Genus") %>%
            taxatree_models(rank = NULL, formula = "~ Days.post.exposure.start +
                            (1|treatment/tank/PIT.tag.num)",
                      type = lmerTest::lmer) %>%
            taxatree_models2stats(fun = broom.mixed::tidy) %>%
            taxatree_stats_get() %>%
            filter(effect == "fixed") %>%
            mutate(p.adj = p.adjust(p.value, method = "BH"),
                   Timepoint = paste0(comp$TP1[TEST],"_",comp$TP2[TEST]),
                   Sample.type = SAM.TYPE,
                   treatment = TREATMENT,
                   iteration = ITER)
            )
        } else if (SAM.TYPE == "Filter") {
          diff.abund <- bind_rows(diff.abund,
                              ps %>%
            tax_fix() %>%
            ps_filter(sample.type == SAM.TYPE,
                      Days.post.exposure.start %in% c(comp$TP1[TEST], comp$TP2[TEST])) %>%
            tax_agg(rank = "Genus") %>%
            tax_prepend_ranks() %>%
            tax_filter(min_sample_abundance = 30, tax_level = "Genus") %>%
            taxatree_models(rank = NULL, formula = "~ Days.post.exposure.start +
                            (1|treatment/tank)",
                      type = lmerTest::lmer) %>%
            taxatree_models2stats(fun = broom.mixed::tidy) %>%
            taxatree_stats_get() %>%
            filter(effect == "fixed") %>%
            mutate(p.adj = p.adjust(p.value, method = "BH"),
                   Timepoint = paste0(comp$TP1[TEST],"_",comp$TP2[TEST]),
                   Sample.type = SAM.TYPE,
                   treatment = TREATMENT,
                   iteration = ITER)
            )
        } else {
          stop("Sample Type not Filter or Swab")
        }
      }
    }
  }
}
(time <- round(Sys.time() - start, digits = 2))

file.list <- Sys.glob("data/diff_abund_*.csv")

diff.abund <- bind_rows(read_csv(file.list))

filt <- diff.abund %>%
  filter(str_detect(term, ":factor")) %>%
  mutate(TP2 = as.integer(str_extract(term, "\\d+$"))) %>%
  filter(Timepoint < TP2) %>%
  mutate(treatment = if_else(str_detect(term, "Low"), "Low",
                             if_else(str_detect(term, "High"), "High", NA))) %>%
  group_by(term, taxon, rank, formula, Timepoint, Sample.type, TP2, treatment) %>%
  summarise(across(where(is.double), mean), .groups = "drop")

sample_size <- round(sample_sums(ps)[1]/1000 * 0.95)

filt.swab <- filt %>%
  #filter(p.adj < 0.05) %>%
  filter(Sample.type == "Swab") %>%
  #mutate(term = str_remove_all(term, "treatment|:Days.post.exposure.start")) %>%
  #unite("term", treatment, term, Timepoint) %>%
  mutate(term = paste0(treatment, "_", Timepoint, "_", TP2)) %>%
  arrange(treatment, Timepoint, TP2) %>%
  mutate(term = factor(term, levels = unique(term)),
         estimate = estimate / sample_size * 100)

filt.filt <- filt %>%
  #filter(p.adj < 0.05) %>%
  filter(Sample.type == "Filter") %>%
  mutate(term = paste0(treatment, "_", Timepoint, "_", TP2)) %>%
  arrange(treatment, Timepoint, TP2) %>%
  mutate(term = factor(term, levels = unique(term)),
         estimate = estimate / sample_size * 100)

ps.swab <- ps %>%
  tax_fix() %>%
  ps_filter(sample.type == "Swab") %>%
  tax_filter(min_sample_abundance = 30,
             tax_level = "Genus",
             min_prevalence = 0.1,
             min_total_abundance = 0.005
             ) %>%
  tax_agg(rank = "Genus") %>%
  tax_transform(trans = "identity", rank = "Genus") %>%
  tax_prepend_ranks() %>%
  tax_select(tax_list = unique(filt.swab$taxon)) %>%
  taxatree_models(rank = NULL, formula = "~ treatment +
                                   (1 | treatment)",
                    type = lmerTest::lmer) %>%
          taxatree_models2stats(fun = broom.mixed::tidy)

ps.filt <- ps %>%
  tax_fix() %>%
  ps_filter(sample.type == "Filter") %>%
  tax_filter(min_sample_abundance = 30,
             tax_level = "Genus",
             min_prevalence = 0.1,
             min_total_abundance = 0.005
             ) %>%
  tax_agg(rank = "Genus") %>%
  tax_transform(trans = "identity", rank = "Genus") %>%
  tax_prepend_ranks() %>%
  tax_select(tax_list = unique(filt.filt$taxon)) %>%
  taxatree_models(rank = NULL, formula = "~ treatment +
                                   (1 | treatment)",
                    type = lmerTest::lmer) %>%
          taxatree_models2stats(fun = broom.mixed::tidy)

ps.swab@taxatree_stats <- filt.swab
ps.filt@taxatree_stats <- filt.filt

rename.title <- function(x) { 
                  str_remove(x, "High_|Low_") %>%
                  str_replace(., "-2", "Pre-exposure") %>%
                  str_replace(., "7", "Exposure 1") %>%
                  str_replace(., "14", "Exposure 2") %>%
                  str_replace(., "21", "Depuration 1") %>%
                  str_replace(., "28", "Depuration 2") %>%
                  str_replace_all(., "_", " versus ")
}

tax.trees.swab <- ps.swab %>%
   taxatree_label(p.adj < 0.05 & rank %in% c("Genus", "Phylum")) %>%
  taxatree_plots(sig_stat = "p.adj",
                colour_trans = "identity",
                palette = "Blue-Red 3", 
                l2 = 90,
                sig_shape = "circle filled", 
                sig_size = 1.5,
                sig_stroke = 1.5,
                sig_colour = "white",
                sig_threshold = 0.05,
                layout = "fr",
                layout_seed = 1234,
                colour_lims = range(filt.swab$estimate),
                var_renamer = rename.title)

for (tree in 1:length(tax.trees.swab)) {
  tax.trees.swab[[tree]] <- tax.trees.swab[[tree]] %>%
    taxatree_plot_labels(
    taxon_renamer = function(x) stringr::str_remove(x, "[PCOFG]: "),
    circular = F,
    fun = ggrepel::geom_text_repel, box.padding = 1, force = 5, seed = 1234)
}

tax.trees.filt <- ps.filt %>%
     taxatree_label(p.adj < 0.05) %>%
  taxatree_plots(sig_stat = "p.adj",
                colour_trans = "identity",
                palette = "Blue-Red 3", 
                l2 = 90,
                sig_shape = "circle filled", 
                sig_size = 1.5,
                sig_stroke = 1.5,
                sig_colour = "white",
                sig_threshold = 0.05,
                layout = "fr",
                layout_seed = 1234,
                colour_lims = range(filt.filt$estimate),
                var_renamer = rename.title)

for (tree in 1:length(tax.trees.filt)) {
  tax.trees.filt[[tree]] <- tax.trees.filt[[tree]] %>%
    taxatree_plot_labels(
    taxon_renamer = function(x) stringr::str_remove(x, "[PCOFG]: "),
    circular = F,
    fun = ggrepel::geom_text_repel, box.padding = 1, force = 5, seed = 1234)
}

rename.legend <- labs(colour = "Estimate\n(Relative abundance)", 
       fill = "Estimate\n(Relative abundance)", 
       edge_colour ="Estimate\n(Relative abundance)",
       size = "Prevalence",
       edge_width = "Prevalence")

prefixes <- c("Low_", "High_")
suffixes <- unite(comp, "ABC", TP1,TP2)$ABC

plots.swab <- list()
for (prefix in prefixes) {
  for (suffix in suffixes) {
    col_name <- paste(prefix, suffix, sep = "")
    plot_name <- paste(prefix, suffix, sep = "")
    
    if (!is.null(tax.trees.swab[[col_name]])) {
      plots.swab[[plot_name]] <- tax.trees.swab[[col_name]] + rename.legend
    } else {
      plots.swab[[plot_name]] <- plot_spacer()
    }
  }
}

plots.filt <- list()
for (prefix in prefixes) {
  for (suffix in suffixes) {
    col_name <- paste(prefix, suffix, sep = "")
    plot_name <- paste(prefix, suffix, sep = "")
    
    if (!is.null(tax.trees.filt[[col_name]])) {
      plots.filt[[plot_name]] <- tax.trees.filt[[col_name]] + rename.legend
    } else {
      plots.filt[[plot_name]] <- plot_spacer()
    }
  }
}

design <- "
ABCD
#EFG
H#IJ
###K
"

A <- plots.swab[grep("Low", names(plots.swab))]
tax.trees.swab.low <- 
  A[[1]] + A[[2]] + A[[3]] + A[[4]] + 
           A[[5]] + A[[6]] + A[[7]] + 
guide_area() +      A[[8]] + A[[9]] + 
                            A[[10]] + 
  plot_layout(guides = 'collect', design = design)

ggsave(plot = tax.trees.swab.low,
       filename = "figures/taxatrees.diff.abund.swab.low.svg",
       width = 30, height = 20, units = "cm")


A <- plots.swab[grep("High", names(plots.swab))]
tax.trees.swab.high <- 
A[[1]] + A[[2]] + A[[3]] + A[[4]] + 
         A[[5]] + A[[6]] + A[[7]] + 
guide_area() +    A[[8]] + A[[9]] + 
                          A[[10]] + 
  plot_layout(guides = 'collect', design = design)

ggsave(plot = tax.trees.swab.high,
       filename = "figures/tax.trees.swab.high.taxatrees.svg",
       width = 30, height = 20, units = "cm")

A <- plots.filt[grep("Low", names(plots.filt))]
tax.trees.filt.low <- 
  A[[1]] + A[[2]] + A[[3]] + A[[4]] + 
           A[[5]] + A[[6]] + A[[7]] + 
guide_area() +      A[[8]] + A[[9]] + 
                            A[[10]] + 
  plot_layout(guides = 'collect', design = design)

ggsave(plot = tax.trees.filt.low,
       filename = "figures/taxatrees.diff.abund.filt.low.svg",
       width = 30, height = 20, units = "cm")

A <- plots.filt[grep("High", names(plots.filt))]
tax.trees.filt.high <- 
A[[1]] + A[[2]] + A[[3]] + A[[4]] + 
         A[[5]] + A[[6]] + A[[7]] + 
guide_area() +    A[[8]] + A[[9]] + 
                          A[[10]] + 
  plot_layout(guides = 'collect', design = design)

ggsave(plot = tax.trees.filt.high,
       filename = "figures/tax.trees.filt.high.taxatrees.svg",
       width = 30, height = 20, units = "cm")
```

# Figure 4
## Alluvial plot most diff abund taxa
```{r}
comp <- c("-2_7", "7_14", "14_21", "21_28")

filt.filt %>%
  select(Timepoint, Timepoint, TP2, estimate, taxon, treatment) %>%
  rename(TP1 = Timepoint) %>%
  unite(Timepoint, TP1, TP2, remove = T) %>%
  filter(Timepoint %in% comp) %>%
  mutate(Timepoint = factor(Timepoint, levels = comp)) %>%
  filter(taxon %in% c("G: Arcicella","G: Cetobacterium","G: Acinetobacter",
                      "G: Flavobacterium","G: Sphaerotilus", "G: Cellvibrio",
                      "G: Roseateles", "G: Flectobacillus", "G: Aeromonas")) %>%
  mutate(taxon = str_replace(taxon, "G: ", "")) %>%
  mutate(relationship = if_else(taxon %in% c("Arcicella","Cetobacterium","Acinetobacter","Flavobacterium"), 
                                "Enriched under antibiotic exposure",
                        if_else(taxon %in% c("Sphaerotilus", "Cellvibrio"), 
                                "Depleted under antibiotic exposure",
                        if_else(taxon %in% c("Roseateles", "Flectobacillus", "Aeromonas"), "Increasing / decreasing throughout", NA)))) %>%
  filter(relationship != "Increasing / decreasing throughout") %>%
  group_by(taxon, treatment) %>%
  mutate(estimate2 = max(estimate) + 0.001 - estimate) %>%
  ungroup() %>%
  mutate(taxon = factor(taxon),
         treatment = factor(treatment),
         relationship = factor(relationship),
         TP = as.numeric(Timepoint)) %>%
  ggplot(aes(x = TP, y = estimate2, alluvium = taxon)) +
  geom_alluvium(aes(fill = taxon, colour = taxon)) +
  #scale_x_continuous(breaks = seq(1, 4, 1)) +
  annotate("rect", xmin = 0, xmax = 2, ymin = -Inf, ymax = Inf, alpha=0.2, fill="black") +
  theme_bw() +
  coord_cartesian(xlim = c(1, 4), ylim = c(0, NA)) +
  facet_grid(treatment~relationship) +
  scale_fill_brewer(palette = "Set1", aesthetics = c("colour", "fill")) +
  scale_x_continuous(labels=c("D-2 vs. 7", "D7 vs. 14", "D14 vs. 21", "D21 vs. 28")) +
  xlab("Time point comparison") +
  ylab("Change in estimate") +
  guides(fill=guide_legend(title= "Genus"), colour = guide_legend(title= "Genus"))

```

# Diff abund stats
```{r}
bind_rows(filt.swab, filt.filt) %>%
  select(term, taxon, formula, estimate, std.error, Sample.type, p.adj) %>%
  #filter(p.adj < 0.05) %>%
  distinct() %>%
  write_csv("data/diff.abund.taxa.csv")
```

# Supplementary Figure 5
# Supplementary table 6
```{r}
comp <- c(-2,7,14,21,28)
comp <- expand.grid(TP1 = comp, TP2 = comp)
comp <- comp[comp$TP1 != comp$TP2 & comp$TP1 < comp$TP2, ] %>%
  mutate(Pre_treatment = if_else(TP1 == -2 | TP2 == -2, T, F),
         Exposure = if_else(TP1 %in% c(7,14) | TP2 %in% c(7,14), T, F),
         Depuration = if_else(TP1 %in% c(21,28) | TP2 %in% c(21,28), T, F)) %>%
  pivot_longer(cols = c(Pre_treatment, Exposure, Depuration)) %>%
  filter(value != F) %>%
  select(-value) %>%
  group_by(TP1, TP2) %>%
  reframe(comparison = paste0(first(name),"_vs_",last(name))) %>%
  unite(Timepoint, TP1, TP2, remove = F)

diff_stats <- function(MATRIX, NAME){
  as.matrix(MATRIX) %>% # convert matrix to matrix
  reshape2::melt(varnames = c("Sample1", "Sample2")) %>% # get into pivot_longer format
  left_join(samdat_tbl(ps), by = join_by(Sample1 == .sample_name)) %>% # get metadata for first comparison
  select(treatment, tank, PIT.tag.num, Days.post.exposure.start, Sample2, value) %>% # keep relevant stuff
  filter(!is.na(PIT.tag.num)) %>% # only get swab data
  left_join(samdat_tbl(ps), by = join_by(Sample2 == .sample_name)) %>% # get metadata for second comparison
  select(matches(".x$|.y$"), value) %>% # only keep metadata selected in Sample1 - identical names
  filter(!is.na(PIT.tag.num.y), # remove swab versus filter comparisons
         PIT.tag.num.x == PIT.tag.num.y, # remove self comparisons
         Days.post.exposure.start.x != Days.post.exposure.start.y) %>% # remove same day comparisons
  select(-c(PIT.tag.num.y, treatment.y, tank.y)) %>% # remove sample2 metadata
  arrange(PIT.tag.num.x, Days.post.exposure.start.x, Days.post.exposure.start.y) %>% # arrange nicely
  unite(TP, Days.post.exposure.start.x, Days.post.exposure.start.y) %>% # make it comparison at timepoints
  filter(TP %in% c("-2_7", "-2_14", "7_21", "7_28", "14_21", "14_28")) %>% # only filter out these comparisons
  mutate(TP = factor(TP, levels = c("-2_7", "-2_14", "7_21", "7_28", "14_21", "14_28"))) %>%
  lmerTest::lmer(value ~ treatment.x*TP + (1 + treatment.x|treatment.x/tank.x/PIT.tag.num.x), data = .) %>%
  broom.mixed::tidy() %>% # test dissimialrity between samples
  filter(str_detect(term, ":")) %>%
  mutate(`B.H. p adjusted` = p.adjust(p.value, method = "BH"), # FDR
         `Differences between time points` = 
           str_replace(term, "treatment.xLow:TP|treatment.xHigh:TP", ""), # get user friendly names
         `Differences between time points` = 
           str_replace(`Differences between time points`, "_", " vs. "),
         term = str_replace(term, ".*High.*", "High"),
         term = str_replace(term, ".*Low.*", "Low"),
         `Beta diversity metric` = NAME) %>% # name of matrix
  rename(Treatment = term) %>%
  select(`Beta diversity metric`, Treatment, `Differences between time points`, estimate, std.error, statistic, df, `B.H. p adjusted`) # keep useful information
}

x <- bind_rows(diff_stats(BC.agg, "Bray-Curtis"), 
          diff_stats(Jacc.agg, "Binary Jaccard"), 
          diff_stats(WUF.agg, "UF"), 
          diff_stats(rclr.agg, "Atitchison"))

x %>%
  select(-std.error, -df, -statistic) %>%
  mutate(across(where(is.double), ~round(., digits = 3))) %>%
  rename(`B.H. FDR` = `B.H. p adjusted`) %>%
  mutate(`B.H. FDR` = if_else(`B.H. FDR` == 0, "p < 0.001 ***", 
                      if_else(`B.H. FDR` < 0.05, paste0(`B.H. FDR`, " *"), as.character(`B.H. FDR`))),
         `Differences between time points` = factor(`Differences between time points`, levels = c("-2 vs. 7", "-2 vs. 14", "7 vs. 21", "7 vs. 28", "14 vs. 21", "14 vs. 28"))) %>%
  pivot_wider(names_from = `Beta diversity metric`, values_from = c(`B.H. FDR`, estimate)) %>%
  arrange(Treatment, `Differences between time points`) %>%
  view()

```

# Figure 5
## ARGs
```{r}
#RGI antibiotics stuff
RGI.files <- Sys.glob("data/Sample*RGI.txt")
RGI <- read_tsv(RGI.files)

cov.stat.files <- Sys.glob("data/Sample*_GC*covstat.txt")
covstat <- read_tsv(cov.stat.files, col_names = c("Contig", "RPKM", "TPM")) %>%
  filter(Contig %in% RGI$ORF_ID)

key <- tibble(Sample = paste0(rep("Sample_", 24), 1:24),
              Plate.no = as.factor(c(5,4,3,3,4,5,5,5,4,1,4,5,5,5,3,4,2,1,5,2,1,3,2,4)),
              well = c("B09","H02","G06","B05","F09","H12","G04","H01","D07",
                       "D11","E10","E05","G12","H07","F07","E08","G01","H09",
                       "D05","D09","F02","A01","F04","D05"))

meta <- samdat_tbl(ps)

df <- left_join(RGI, covstat, by = join_by(ORF_ID == Contig)) %>%
  mutate(Sample = str_split_i(ORF_ID, "_NODE_", 1))%>%
  left_join(key, by = "Sample") %>%
  left_join(meta, by = c("Plate.no", "well"))
# check
df %>%
  select(sample.type, Sample, Tank) %>%
  unique() %>%
 print(n = 24)

write_csv(df, "data/ARG_data.csv")

ARG.expand <- ARG %>%
  mutate(Drug.Class.Expand = strsplit(as.character(`Drug Class`), "; ")) %>%
  unnest(Drug.Class.Expand) %>%
  mutate(Drug.Class.Expand = str_trim(gsub("antibiotic", "", Drug.Class.Expand), side = c("both")))

ARG.strict <- ARG.expand %>%
  filter(Cut_Off == "Strict")

AB.int <- c("sulfonamide", "diaminopyrimidine", "fluoroquinolone",
                                  "macrolide", "tetracycline")

length(unique(ARG$Best_Hit_ARO))
length(unique(ARG.expand$Drug.Class.Expand))

ARG.expand %>%
  select(sample.type, treatment, Drug.Class.Expand, RPKM) %>%
  group_by(Drug.Class.Expand) %>%
  summarise(sum.RPKM = sum(RPKM), .groups = "drop") %>%
  arrange(desc(sum.RPKM)) %>% 
  print(n = 100)

ARG.plot <- function(COMPARISON) {
  ARG.expand %>%
  filter(Drug.Class.Expand %in% AB.int,
         treatment %in% c("Control", COMPARISON)) %>%
  group_by(sample.type, treatment, Drug.Class.Expand, Sample, replicate) %>%
  summarise(sum.RPKM = sum(RPKM), .groups = "drop") %>%
  unite(Sam_drug, sample.type, Drug.Class.Expand, remove = F) %>%
  unite(Drug_rep, Drug.Class.Expand, replicate, remove = F) %>%
  ggplot(aes(x=treatment, y=sum.RPKM)) +
  geom_point(aes(colour=Drug_rep), size = 2) +
  geom_line(stat = "smooth", aes(group = replicate, colour = Drug_rep),
              method = "lm", se = F, linewidth = 1, linetype = "dashed") +
  geom_line(stat = "smooth", aes(group = Sam_drug), colour = "black",
              method = "lm", se = F, linewidth = 1) +
  ggh4x::facet_grid2(Drug.Class.Expand~sample.type, scales = "free", independent = "all") +
  theme_bw() +
  scale_color_brewer(palette = "Paired") +
  theme(legend.position = "none") +
  xlab("Treatment group") +
  ylab("Sum of ARG RPKMs by sample")
}

plot1 <- ARG.plot("Low") + ARG.plot("High") + plot_layout(axis_titles = "collect") + plot_annotation(tag_level = "A", caption = "Linear mixed model with random effects show all test not significant")

ggsave(plot = plot1, filename = "figures/ARG.abundance.by.treatment.sam.type.svg", width = 25.5*1.25, height = 15.5*1.25, units = "cm")


clean.stats <- function(x){
  x %>%
    filter(effect == "fixed",
         term != "(Intercept)") %>%
  mutate(treatment = if_else(str_detect(term, "treatmentLow"), "Low", "High"),
         tank = if_else(str_detect(term, "tankHighA"), "HighA",
                        if_else(str_detect(term, "tankHighC"), "HighC",
                                if_else(str_detect(term, "tankLowA"), "LowA",
                                        if_else(str_detect(term, "tankLowC"), "LowC",
                                                if_else(str_detect(term, "tankControlC"), "ControlC", "ControlA"))))),
         p.adj = p.adjust(p.value, method = "BH"),
    Significance = if_else(p.adj <= 0.001, "***",
                            if_else(p.adj <= 0.01, "**",
                                    if_else(p.adj <= 0.05, "*","N.S."))),
    across(where(is.double), ~ round(., digits = 2)),
    p.adj = if_else(p.adj == 0, as.character("p < 0.001"), as.character(p.adj)))
}

posslm1 = possibly(.f = lmerTest::lmer, otherwise = "Error", quiet = T)
poss.broom = possibly(.f = broom.mixed::tidy, otherwise = NA, quiet = T)

ARG.test.drug.class <- ARG.expand %>%
  select(sample.type, treatment, tank, Drug.Class.Expand, RPKM, Sample, replicate) %>%
  group_by(sample.type, treatment, tank, Drug.Class.Expand, Sample, replicate) %>%
  summarise(sum.RPKM = sum(RPKM), .groups = "drop") %>%
  nest(ARG.data = c(treatment, tank, replicate, Sample, sum.RPKM)) %>%
  arrange(sample.type, Drug.Class.Expand) %>%
  mutate(ARG.lmer = map(ARG.data, function(x)
    posslm1(sum.RPKM ~ treatment + tank + (1 | replicate/tank) + (1 | treatment/tank), 
               data = x, na.action = na.omit))) 

ARG.test.drug.class %>%
  mutate(ARG.lmer = map(ARG.lmer, function(x)
    poss.broom(x))) %>%
    select(-ARG.data) %>%
    unnest(ARG.lmer) %>%
  clean.stats(.) %>%
  #filter(Drug.Class.Expand %in% AB.int) %>%
  filter(p.value < 0.05)
  view()

#### Individual ARGs

ARG.test2 <- ARG %>%
  select(sample.type, treatment, tank, replicate, Sample, Best_Hit_ARO, RPKM) %>%
  group_by(sample.type, treatment, tank, replicate, Sample, Best_Hit_ARO) %>%
  summarise(sum.RPKM = sum(RPKM), .groups = "drop") %>%
  nest(ARG.data = c(treatment, tank, replicate, Sample, sum.RPKM)) %>%
  arrange(sample.type, Best_Hit_ARO) %>%
  mutate(ARG.lmer = map(ARG.data, function(x)
    posslm1(sum.RPKM ~ treatment + tank + (1 | replicate/tank) + (1 | treatment/tank), 
               data = x, na.action = na.omit) %>%
  poss.broom(.))) %>%
  select(-ARG.data) %>%
  unnest(ARG.lmer) %>%
  clean.stats()

ARG.test2 %>%
  filter(p.adj < 0.05)

ARG.res <- ARG.expand %>%
  select(Best_Hit_ARO, Drug.Class.Expand, `Drug Class`) %>%
  distinct() %>%
  right_join(ARG.test2, by = "Best_Hit_ARO", relationship = "many-to-many")

ARG.filt <- ARG.res %>% 
  filter(str_detect(`Drug Class`, paste(AB.int, collapse = "|"))) %>%
  select(-Drug.Class.Expand) %>%
  distinct()

# ref length 60 (87.5), perc ID 75
# neg controls

ggplot(ARG, aes(x = Best_Identities)) +
  geom_histogram(binwidth = 1) +
  xlim(20, 50)

ggplot(ARG, aes(x = `Percentage Length of Reference Sequence`)) +
  geom_histogram(binwidth = 1) +
  xlim(80, 120)

quantile(ARG$Best_Identities, probs = c(0.01, 0.05, 0.1, 0.25))
quantile(ARG$`Percentage Length of Reference Sequence`, probs = c(0.01, 0.05, 0.1, 0.25))

num.genes <- ARG %>%
  group_by(sample.type, Sample) %>%
  summarise(n = n()) %>%
  group_by(sample.type) %>%
  summarise(min = min(n))

swab_data <- ARG %>%
  filter(sample.type == "Swab") %>%
  select(Sample, sample.type, treatment, tank, Drug.Class.Expand, RPKM)

filter_data <- ARG %>%
  filter(sample.type == "Filter") %>%
  select(Sample, sample.type, treatment, tank, Drug.Class.Expand, RPKM)

data <- bind_rows(
  lapply(1:10, function(i) {
    resample.swab <- swab_data %>%
      group_by(Sample) %>%
      sample_n(size = round(num.genes$min[num.genes$sample.type == "Swab"] * 0.95))

    resample.filter <- filter_data %>%
      group_by(Sample) %>%
      sample_n(size = round(num.genes$min[num.genes$sample.type == "Filter"] * 0.95))

    bind_rows(resample.swab, resample.filter)
  })
)

data %>%
  filter(Drug.Class.Expand %in% AB.int) %>%
  group_by(sample.type, treatment, tank, Drug.Class.Expand) %>%
  summarise(sum.RPKM = sum(RPKM), count = n(), .groups = "drop") %>%
ggplot(aes(x=treatment, y=sum.RPKM)) +
  geom_point(aes(colour = Drug.Class.Expand)) +
  #geom_boxplot(aes(fill=Drug.Class.Expand)) +
  stat_summary(fun = median, geom = "line", aes(group = "treatment"), linewidth = 1) +
  ggh4x::facet_grid2(Drug.Class.Expand~sample.type, scales = "free", independent = "all") +
  theme_bw()
```
```{r}
Table 1
AB <- read_csv("data/water.chemistry_sample.tubes.csv") %>%
  mutate(across(.cols = everything(), ~str_replace(., "< MQL|< MDL|< IQL|< IDL|#N/A|No ISTD", "0"))) %>%
  replace(is.na(.), "0")

AB.colnames <- colnames(AB)[grepl("µg/L", colnames(AB))]

AB[AB.colnames] <- sapply((AB[AB.colnames]), as.numeric)

AB.filt <- AB %>%
  select(Days.post.exposure.start, Tank, AB.concentration.group,
         all_of(AB.colnames)) %>%
  filter(AB.concentration.group != "Matrix QC") %>%
  mutate(Days.post.exposure.start = as.numeric(Days.post.exposure.start)) %>%
  pivot_longer(cols = all_of(AB.colnames),
               names_to = "Antibiotic_Group", values_to = "Antibiotic_Concentration") %>%
  separate(Antibiotic_Group, c("target.actual","target.antibiotic", NA), sep = "\\.") %>%
  filter(target.antibiotic != "desethylene_ciprofloxacin", target.antibiotic != "desmethyl_clarithromycin") %>%
  mutate(AB.concentration.group = if_else(AB.concentration.group == "high.dose", "High dose exposure", 
                                          if_else(AB.concentration.group == "low.dose", "Low dose exposure", "Negative control")),
         target.antibiotic = str_to_title(target.antibiotic))

AB.filt %>%
    mutate(Antibiotic_Concentration = replace_na(Antibiotic_Concentration, 0)) %>%
  filter(Days.post.exposure.start %in% c(0.25, 1, 2, 7)) %>% # only take were chemistry was measured
  group_by(AB.concentration.group, target.actual, target.antibiotic) %>%
  summarise(meanAB.each = mean(Antibiotic_Concentration),
            SD.AB.each = sd(Antibiotic_Concentration),
            .groups = "drop_last") %>%
  pivot_wider(names_from = target.actual, values_from = c(meanAB.each, SD.AB.each)) %>% 
  write_csv("data/exposure_concentrations.csv")
  
  summarise(meanAB = mean(meanAB.each),
            mean.SD.AB = mean(SD.AB.each),
            .groups = "drop_last") %>%
  pivot_wider(names_from = target.actual, values_from = c(meanAB, mean.SD.AB)) %>%
  mutate(nominal = meanAB_actual/meanAB_target) %>%
  ungroup(Days.post.exposure.start) %>%
  summarise(meanNominal = mean(nominal),
            meanSD = mean(mean.SD.AB_actual))
```
# Figure 6 and Supplementary figure 8
```{r}
humann <- read_tsv("data/all_sample_genefamilies.tsv") %>%
  rename(Gene_Family = `# Gene Family`) %>%
  mutate(pathway = str_extract(Gene_Family, "\\[[0-9].*[0-9]\\]") %>% gsub('^.|.$', '', .),
         pathname = str_extract(Gene_Family, "\\(expasy\\).*\\[") %>% gsub("\\(expasy\\) |\\[$", "", .))

humann.expand <- humann %>%
  filter(!is.na(pathway)) %>%
  group_by(pathway) %>% 
  mutate(pathway_unique = strsplit(pathway, "; ")) %>% 
  unnest(pathway_unique) %>% 
  ungroup(pathway) %>% 
  separate(pathway_unique, c("path1", "path2", "path3", "path4"), "\\.") %>%
  mutate(path2 = paste0(path1, ".", path2),
         path3 = paste0(path2, ".", path3),
         path4 = paste0(path3, ".", path4))

path4 <- humann.expand %>%
  group_by(path4) %>%
  summarise(across(where(is.double), sum)) %>%
  rename_all(~str_remove_all(., ".ec_Abundance-CPM|Sample_")) %>%
  column_to_rownames(var = "path4")

path4.swab <- path4 %>%
  select("1","2","3","4","10","11","12","17","18","19","20")

path4.filt <- path4 %>%
  select("5","6","7","8","13","14","15","16","21","22","23","24")

metadata <- ARG %>% 
  mutate(Sample = as.numeric(str_replace_all(Sample, "Sample_", "")),
         sample.type = as.factor(sample.type)) %>%
  arrange(Sample) %>%
  mutate(Dosage = str_split_i(tank, "[A-Z]$", 1)) %>%
  filter(!Sample == "9")

matrix.swab <- t(path4.swab)[ order(as.numeric(row.names(t(path4.swab)))), ]
matrix.filt <- t(path4.filt)[ order(as.numeric(row.names(t(path4.filt)))), ]

nMDS.swab <- metaMDS(vegdist(matrix.swab), trymax = 1000)
nMDS.filt <- metaMDS(vegdist(matrix.filt), trymax = 1000)

swab <- data.frame(nMDS.swab$points) %>%
  rownames_to_column(var = "Sample") %>%
  mutate(Sample = as.numeric(Sample)) %>%
  left_join(metadata, by = join_by(Sample))

filt <- data.frame(nMDS.filt$points) %>%
  rownames_to_column(var = "Sample") %>%
  mutate(Sample = as.numeric(Sample)) %>%
  left_join(metadata, by = join_by(Sample))

adonis.swab <- pairwise.adonis2(vegdist(matrix.swab) ~ Dosage, data = filter(metadata, sample.type == "Swab"))

adonis.filt <- pairwise.adonis2(vegdist(matrix.filt) ~ Dosage, data = filter(metadata, sample.type == "Filter"))

adonis.filt$High_vs_Control$R2[1]
adonis.filt$Low_vs_Control$R2[1]

(swab.p <- ggplot(swab, aes(MDS1, MDS2)) +
  geom_point(aes(colour = Dosage), size = 4) +
  ggpubr::stat_chull(aes(fill = Dosage, color = Dosage), alpha = 0.1, geom = "polygon") +
  scale_color_manual(values = clrs.tanks, aesthetics = c("colour", "fill")) +
  facet_wrap(~sample.type) +
  annotate(geom="label", 
           label = paste("Pairwise PERMANOVA:\nControl vs Low dose R² = ",
                         round(adonis.swab$Low_vs_Control$R2[1], 3),
                         ", p = ", adonis.swab$Low_vs_Control$`Pr(>F)`[1], 
                         "\nControl vs High dose R² = ",
                         round(adonis.swab$High_vs_Control$R2[1], 3),
                         ", p = ", adonis.swab$High_vs_Control$`Pr(>F)`[1],
                         sep = ""), 
           x=-Inf, y=Inf, hjust = 0, vjust = 1) +
  guides(colour=guide_legend(title = "Treatment group"),
         fill=guide_legend(title = "Treatment group")) +
  theme_bw())

(filt.p <- ggplot(filt, aes(MDS1, MDS2)) +
  geom_point(aes(colour = Dosage), size = 4) +
  ggpubr::stat_chull(aes(fill = Dosage, color = Dosage), alpha = 0.1, geom = "polygon") +
  scale_color_manual(values = clrs.tanks, aesthetics = c("colour", "fill")) +
  facet_wrap(~sample.type) +
  annotate(geom="label", 
           label = paste("Pairwise PERMANOVA:\nControl vs Low dose R² = ",
                         round(adonis.filt$Low_vs_Control$R2[1], 3),
                         ", p = ", adonis.filt$Low_vs_Control$`Pr(>F)`[1], 
                         "\nControl vs High dose R² = ",
                         round(adonis.filt$High_vs_Control$R2[1], 3),
                         ", p = ", adonis.filt$High_vs_Control$`Pr(>F)`[1],
                         sep = ""), 
           x=-Inf, y=Inf, hjust = 0, vjust = 1) +
  guides(colour=guide_legend(title = "Treatment group"),
         fill=guide_legend(title = "Treatment group")) +
  theme_bw())

swab.p + filt.p +  patchwork::plot_annotation(tag_levels = 'A') +
  plot_layout(guides = "collect")

df <- t(path4.filt) %>%
  as.data.frame() %>%
  rownames_to_column(var = "Sample") %>%
  tibble() %>%
  mutate(Sample = as.double(Sample)) %>%
  left_join(metadata, by = join_by(Sample))

df.low.ctrl <- df %>%
  filter(!Dosage == "High") %>%
  column_to_rownames(var = "Sample") %>%
  select(-tank, -Dosage, -sample.type) %>%
  select(which(!colSums(., na.rm=TRUE) %in% 0)) %>%
  rownames_to_column(var = "Sample") %>%
  mutate(Sample = as.integer(Sample)) %>%
  left_join(metadata, by = join_by("Sample"))

df.low.ctrl.meta <- df.low.ctrl %>%
  select(Sample, tank, Dosage)
  
collect <- data.frame()
for (column in 2:(ncol(df.low.ctrl)-3)) {
  tmp <- summary(lmerTest::lmer(df.low.ctrl[[column]] ~ Dosage + (1|tank), data = df.low.ctrl.meta, na.action = na.omit))
  tmp2 <- data.frame(pathway = colnames(df.low.ctrl[column]), 
                     p.value = as.data.frame(tmp$coefficients)$`Pr(>|t|)`[2])
  collect <- rbind(collect, tmp2)
}

sig.pathways <- collect %>%
  mutate(p.adj = p.adjust(p.value, method = "BH")) %>%
  filter(p.adj < 0.05)

key <- humann.expand %>%
  select(path4, pathname) %>%
  unique()

(low.dose.p <- df.low.ctrl %>%
  select(Sample, Dosage, tank, sig.pathways$pathway) %>%
  pivot_longer(cols = c(-Sample, -Dosage, -tank), names_to = "pathway", values_to = "CPM") %>%
  left_join(key, by = join_by(pathway == path4)) %>%
  mutate(pathname = str_split_i(pathname, " $", 1),
         combname = paste0(pathname, "\n", pathway)) %>%
  mutate(Sample = as.factor(Sample)) %>%
  ggplot(aes(x = CPM, y = combname)) +
  geom_point(aes(colour = Dosage), size = 4) +
  scale_colour_manual(values = clrs.tanks) +
  ylab("MetaCyc Pathway") +
  xlab("Copies per Million") +
  guides(colour=guide_legend(title = "Treatment group")) +
  labs(caption = "All differentially abundant pathways determined by linear mixed model 'lmer(CPM ~ treatment.group + (1|tank))' corrected for FDR, p < 0.05") +
  theme_bw())

### High

df.high.ctrl <- df %>%
  filter(!Dosage == "Low") %>%
  column_to_rownames(var = "Sample") %>%
  select(-tank, -Dosage, -sample.type) %>%
  select(which(!colSums(., na.rm=TRUE) %in% 0)) %>%
  rownames_to_column(var = "Sample") %>%
  mutate(Sample = as.integer(Sample)) %>%
  left_join(metadata, by = join_by("Sample"))

df.high.ctrl.meta <- df.high.ctrl %>%
  select(Sample, tank, Dosage)
  
collect.high <- data.frame()
for (column in 2:(ncol(df.high.ctrl)-3)) {
  tmp <- summary(lmerTest::lmer(df.high.ctrl[[column]] ~ Dosage + (1|tank), data = df.high.ctrl.meta, na.action = na.omit))
  tmp2 <- data.frame(pathway = colnames(df.high.ctrl[column]), 
                     p.value = as.data.frame(tmp$coefficients)$`Pr(>|t|)`[2])
  collect.high <- rbind(collect.high, tmp2)
}

sig.pathways.high <- collect.high %>%
  mutate(p.adj = p.adjust(p.value, method = "BH")) %>%
  filter(p.adj < 0.05)

key <- humann.expand %>%
  select(path4, pathname) %>%
  unique()

(high.dose.p <- df.high.ctrl %>%
  select(Sample, Dosage, tank, sig.pathways.high$pathway) %>%
  pivot_longer(cols = c(-Sample, -Dosage, -tank), names_to = "pathway", values_to = "CPM") %>%
  left_join(key, by = join_by(pathway == path4)) %>%
  mutate(pathname = str_split_i(pathname, " $", 1),
         combname = paste0(pathname, "\n", pathway)) %>%
  mutate(Sample = as.factor(Sample)) %>%
  ggplot(aes(x = CPM, y = combname)) +
  geom_point(aes(colour = Dosage), size = 4) +
  scale_colour_manual(values = clrs.tanks) +
  ylab("MetaCyc Pathway") +
  xlab("Copies per Million") +
  guides(colour=guide_legend(title = "Treatment group")) +
  labs(caption = "All differentially abundant pathways determined by linear mixed model 'lmer(CPM ~ treatment.group + (1|tank))' corrected for FDR, p < 0.05") +
  theme_bw())

### comb
common.paths <- inner_join(sig.pathways, sig.pathways.high, by = join_by(pathway)) %>%
  select(pathway)

df %>%
  filter(sample.type == "Filter") %>%
  column_to_rownames(var = "Sample") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "pathway") %>%
  filter(pathway %in% common.paths$pathway) %>%
  column_to_rownames(var = "pathway") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "Sample") %>%
  mutate(Sample = as.double(Sample)) %>%
  left_join(metadata, by = join_by(Sample)) %>%
  pivot_longer(cols = c(-Sample, -Dosage, -tank, -sample.type), names_to = "pathway", values_to = "CPM") %>%
  left_join(key, by = join_by(pathway == path4)) %>%
  mutate(pathname = str_split_i(pathname, " $", 1),
         combname = paste0(pathname, "\n", pathway)) %>%
  mutate(Sample = as.factor(Sample),
         CPM = as.double(CPM),
         Dosage = as.factor(Dosage)) %>%
  ggplot(aes(x = CPM, y = combname)) +
  geom_jitter(aes(colour = Dosage), size = 4, height = 0.1) +
  scale_colour_manual(values = clrs.tanks) +
  ylab("MetaCyc Pathway") +
  xlab("Copies per Million") +
  guides(colour=guide_legend(title = "Treatment group")) +
  labs(caption = "All differentially abundant pathways determined by linear mixed model against the control \n'lmer(CPM ~ treatment.group + (1|tank))' corrected for FDR, p < 0.05") +
  theme_bw()
```
